crates/holochain_p2p/src/spawn/actor.rs

```diff
@@ -192,18 +198,176 @@ impl event::HcP2pHandler for WrapEvtSender {
+                    }
+                    CallRemoteRes { msg_id, .. } => {
+                        if let Some(resp) = pending.lock().unwrap().respond(msg_id) {
+                            let _ = resp.send(msg);
+                        }
+                    }
+                    RemoteSignalEvt {
+                        to_agent,
+                        zome_call_params_serialized,
+                        signature,
+                    } => {
+                        let dna_hash = DnaHash::from_k2_space(&space);
+                        // remote signals are fire-and-forget
+                        // so it's safe to ignore the response
+                        let _response = evt_sender
+                            .call_remote(dna_hash, to_agent, zome_call_params_serialized, signature)
+                            .await;
+                    }
+                }
+                HolochainP2pResult::Ok(())
+            });
+        }
+        Ok(())
+    }
+}
+
+impl kitsune2_api::KitsuneHandler for HolochainP2pActor {
+    fn create_space(
+        &self,
+        _space: kitsune2_api::SpaceId,
+    ) -> BoxFut<'_, kitsune2_api::K2Result<kitsune2_api::DynSpaceHandler>> {
+        Box::pin(async move {
+            let this: Weak<dyn kitsune2_api::SpaceHandler> = self.this.clone();
+            if let Some(this) = this.upgrade() {
+                Ok(this)
+            } else {
+                Err(kitsune2_api::K2Error::other("instance has been dropped"))
+            }
+        })
+    }
 }
 
 impl HolochainP2pActor {
     /// constructor
     pub async fn create(
-        _handler: event::DynHcP2pHandler,
-        _compat: NetworkCompatParams,
+        config: HolochainP2pConfig,
+        db_peer_meta: DbWrite<DbKindPeerMetaStore>,
+        db_op: DbWrite<DbKindDht>,
+        handler: event::DynHcP2pHandler,
+        lair_client: holochain_keystore::MetaLairClient,
     ) -> HolochainP2pResult<actor::DynHcP2p> {
-        todo!()
+        let mut builder = if config.k2_test_builder {
+            kitsune2_core::default_test_builder()
+        } else {
+            kitsune2::default_builder()
+        };
+
+        builder.peer_meta_store = Arc::new(HolochainPeerMetaStoreFactory { db: db_peer_meta });
+        builder.op_store = Arc::new(HolochainOpStoreFactory {
+            db: db_op,
+            handler: handler.clone(),
+        });
+
+        let builder = builder.with_default_config()?;
+
+        let pending = Arc::new_cyclic(|this| {
+            Mutex::new(Pending {
+                this: this.clone(),
+                map: HashMap::new(),
+            })
+        });
+
+        let this = Arc::new_cyclic(|this| Self {
+            this: this.clone(),
+            evt_sender: WrapEvtSender(handler),
+            lair_client,
+            kitsune: Mutex::new(None),
+            pending,
+        });
+
+        let kitsune = builder.build(this.clone()).await?;
+
+        *this.kitsune.lock().unwrap() = Some(kitsune);
```

Comment:
There's an annoying chicken-egg problem with using ourselves as the handler. I've solved that through this mutex and the "kitsune" helper function below which extracts this instance (or errors if unset). Happy to entertain other solutions to this problem.

Reply:
Yes, for sure we can have separate types with shared `Arc<Mutex<..` internals. That adds additional boilerplate to remove this slightly odd mechanic. Preference?

Reply:
Actually, I take it back. Making the types separate doesn't remove the need for the mutex around the kitsune instance, since the handler receiving incoming messages by calling send on the kitsune instance.

I'm going to leave things as they are for now. We can always do some refactor in the future involving channels and a message handler task if we want to get away from this current setup.